src/Control/Effect/Fail.hs

```diff
@@ -1,28 +1,49 @@
-{-# LANGUAGE FlexibleInstances, MultiParamTypeClasses, TypeOperators, UndecidableInstances #-}
+{-# LANGUAGE DeriveFunctor, FlexibleInstances, GeneralizedNewtypeDeriving, KindSignatures, MultiParamTypeClasses, TypeOperators, UndecidableInstances #-}
 module Control.Effect.Fail
 ( Fail(..)
 , MonadFail(..)
 , runFail
 , FailC(..)
 ) where
 
+import Control.Applicative (Alternative(..))
 import Control.Effect.Carrier
-import Control.Effect.Fail.Internal
-import Control.Effect.Internal
+import Control.Effect.Error
 import Control.Effect.Sum
+import Control.Monad (MonadPlus(..))
 import Control.Monad.Fail
+import Control.Monad.IO.Class
+import Control.Monad.Trans.Class
+import Data.Coerce
+import Prelude hiding (fail)
+
+newtype Fail (m :: * -> *) k = Fail String
+  deriving (Functor)
+
+instance HFunctor Fail where
+  hmap _ = coerce
+  {-# INLINE hmap #-}
+
+instance Effect Fail where
+  handle _ _ = coerce
+  {-# INLINE handle #-}
+
 
 -- | Run a 'Fail' effect, returning failure messages in 'Left' and successful computations’ results in 'Right'.
 --
 --   prop> run (runFail (pure a)) == Right a
-runFail :: (Carrier sig m, Effect sig) => Eff (FailC m) a -> m (Either String a)
-runFail = runFailC . interpret
+runFail :: FailC m a -> m (Either String a)
+runFail = runError . runFailC
+
+newtype FailC m a = FailC { runFailC :: ErrorC String m a }
+  deriving (Alternative, Applicative, Functor, Monad, MonadIO, MonadPlus, MonadTrans)
 
-newtype FailC m a = FailC { runFailC :: m (Either String a) }
+instance (Carrier sig m, Effect sig) => MonadFail (FailC m) where
+  fail s = FailC (throwError s)
```

Comment:
This is kind of an interesting detail: we’re defining the semantics in the `MonadFail` instance, and then calling into this instance from the `Carrier` instance. This might be a tiny optimization for users of the `MonadFail` interface, while still providing the effect interface, and maintaining the same semantics for each.